packages/wdio-utils/src/test-framework/types.ts

```diff
@@ -0,0 +1,47 @@
+import type { executeHooksWithArgs, executeAsync, runSync } from '../shim'
+
+export type HookFnArgs<T> = (ctx: T) => [unknown, T]
+
+export interface WrapperMethods {
+    executeHooksWithArgs: typeof executeHooksWithArgs
+    executeAsync: typeof executeAsync
+    runSync: typeof runSync
+}
+
+export interface SpecFunction {
+    specFn: Function
+    specFnArgs: any[]
+}
+
+export interface BeforeHookParam<T> {
+    beforeFn: Function
+    beforeFnArgs: HookFnArgs<T>
+}
+
+export interface AfterHookParam<T> {
+    afterFn: Function
+    afterFnArgs: HookFnArgs<T>
+}
+
+export interface JasmineContext {
+    failedExpectations: Array<Record<string, unknown>>
+}
+
+export type SpecArguments = (
+    /**
+     * e.g. before(() => { ... })
+     */
+    [Function] |
+    /**
+     * e.g. before(() => { ... }, 3)
+     */
+    [Function, number] |
+    /**
+     * e.g. it('is a test', () => { ... })
+     */
+    [string, Function] |
+    /**
+     * e.g. it('is a test', () => { ... }, 3)
+     */
+    [string, Function, number]
```

Comment:
Let's make a interface where the number is optional for the hooks and specs

Reply:
what number?

Reply:
having `[string, Function] |` in line 42 makes this optional